sdk/testing/stepwise/environments/docker/environment.go

```diff
@@ -0,0 +1,851 @@
+		Subject: pkix.Name{
+			CommonName: n.Name(),
+		},
+		DNSNames:    []string{"localhost", n.Name()},
+		IPAddresses: []net.IP{net.IPv6loopback, net.ParseIP("127.0.0.1")},
+		ExtKeyUsage: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageServerAuth,
+			x509.ExtKeyUsageClientAuth,
+		},
+		KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment | x509.KeyUsageKeyAgreement,
+		SerialNumber: big.NewInt(mathrand.Int63()),
+		NotBefore:    time.Now().Add(-30 * time.Second),
+		NotAfter:     time.Now().Add(262980 * time.Hour),
+	}
+	n.ServerCertBytes, err = x509.CreateCertificate(rand.Reader, certTemplate, n.Cluster.CACert, n.ServerKey.Public(), n.Cluster.CAKey)
+	if err != nil {
+		return err
+	}
+	n.ServerCert, err = x509.ParseCertificate(n.ServerCertBytes)
+	if err != nil {
+		return err
+	}
+	n.ServerCertPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: n.ServerCertBytes,
+	})
+
+	marshaledKey, err := x509.MarshalECPrivateKey(n.ServerKey)
+	if err != nil {
+		return err
+	}
+	n.ServerKeyPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "EC PRIVATE KEY",
+		Bytes: marshaledKey,
+	})
+
+	n.ServerCertPEMFile = filepath.Join(n.WorkDir, "cert.pem")
+	err = ioutil.WriteFile(n.ServerCertPEMFile, n.ServerCertPEM, 0755)
+	if err != nil {
+		return err
+	}
+
+	n.ServerKeyPEMFile = filepath.Join(n.WorkDir, "key.pem")
+	err = ioutil.WriteFile(n.ServerKeyPEMFile, n.ServerKeyPEM, 0755)
+	if err != nil {
+		return err
+	}
+
+	tlsCert, err := tls.X509KeyPair(n.ServerCertPEM, n.ServerKeyPEM)
+	if err != nil {
+		return err
+	}
+
+	certGetter := reloadutil.NewCertificateGetter(n.ServerCertPEMFile, n.ServerKeyPEMFile, "")
+	if err := certGetter.Reload(); err != nil {
+		return err
+	}
+	tlsConfig := &tls.Config{
+		Certificates:   []tls.Certificate{tlsCert},
+		RootCAs:        n.Cluster.RootCAs,
+		ClientCAs:      n.Cluster.RootCAs,
+		ClientAuth:     tls.RequestClientCert,
+		NextProtos:     []string{"h2", "http/1.1"},
+		GetCertificate: certGetter.GetCertificate,
+	}
+
+	if n.Cluster.clientAuthRequired {
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	}
+	n.TLSConfig = tlsConfig
+
+	return nil
+}
+
+// DockerClusterNode represents a single instance of Vault in a cluster
+type DockerClusterNode struct {
+	NodeID            string
+	Address           *net.TCPAddr
+	HostPort          string
+	Client            *api.Client
+	ServerCert        *x509.Certificate
+	ServerCertBytes   []byte
+	ServerCertPEM     []byte
+	ServerCertPEMFile string
+	ServerKey         *ecdsa.PrivateKey
+	ServerKeyPEM      []byte
+	ServerKeyPEMFile  string
+	TLSConfig         *tls.Config
+	WorkDir           string
+	Cluster           *DockerCluster
+	container         *types.ContainerJSON
+	dockerAPI         *docker.Client
+}
+
+// CreateAPIClient creates and configures a Vault API client to communicate with
+// the running Vault Cluster for this DockerClusterNode
+func (n *DockerClusterNode) CreateAPIClient() (*api.Client, error) {
```

Comment:
`DockerClusterNode` should maybe store this client internally and provide a method to get this client it if one has already been previously created.

Reply:
Deferring and tracking this internally, to follow up in the coming weeks